step/src/main/java/com/google/sps/DatabaseHandler.java

```diff
@@ -15,43 +15,135 @@
   // Method for adding a user to the database
-  public static void addUserToDatabase(User user) {
-    idMap.put(user.getId(), user);    
+  public static void addUser(String firstName, String lastName, int dayBirth,
+    int monthBirth, int yearBirth, String email, String id) {
+    Entity entity = new Entity("User", id);
+    entity.setProperty("firstName", firstName);
+    entity.setProperty("lastName", lastName);
+    entity.setProperty("dayBirth", dayBirth);
+    entity.setProperty("monthBirth", monthBirth);
+    entity.setProperty("yearBirth", yearBirth);
+    entity.setProperty("email", email);
+    entity.setProperty("id", id);
+    datastore.put(entity);    
   }
+ 
+  // Adding a user's notificaton to the databse
+  public static void addNotification(String firstId, String secondId,
+      long timestamp, String type) {
+    Entity entity = new Entity("Notification");
+    entity.setProperty("userId", firstId);
+    entity.setProperty("otherUserId", secondId);
+    entity.setProperty("timestamp", timestamp);
+    entity.setProperty("type", type);
+    datastore.put(entity); 
+  }
+
+  // Getting a user's notifications using their id
+  public static Collection<Notification> getUserNotifications(String id) {
+    Collection<Notification> notifications = new ArrayList<>();
+
+    Filter idFilter = new FilterPredicate("userId", FilterOperator.EQUAL, id);
+
+    // Querying all of the user's notifications from the Datastore.
+    Query query = new Query("Notification").setFilter(idFilter)
+      .addSort("timestamp",SortDirection.DESCENDING);
+
+    PreparedQuery results = datastore.prepare(query);
+
+    // Populating the list
+    for (Entity entity : results.asIterable()) {
+      String firstId = (String) entity.getProperty("userId");
+      String secondId = (String) entity.getProperty("otherUserId");
+      long timestamp = (long) entity.getProperty("timestamp");
+      String notificationType = (String) entity.getProperty("type");
+       
+      Notification notification;
+
+      if (notificationType.equals("matching")) {
+        notification = new MatchNotification(firstId, secondId, timestamp);
+      } else {
+        notification = new MessageNotification(firstId, secondId, timestamp);
+      }
+
+      notifications.add(notification);
+    }
 
-  // Method for getting a user object using an id
-  public static User getUserById(long id) throws NoSuchElementException {
-    if (idMap.containsKey(id)) {
-      return idMap.get(id);    
-    } else {
-      throw new NoSuchElementException();  
-    }    
+    return notifications;    
   }
 
-  // Method for removing a user from our database
-  public static void removeUserById(long id) {
-    if (idMap.containsKey(id)) {
-      idMap.remove(id);    
+  // Method for getting a user's email using their id
+  public static String getUserEmail(String id) {
+    try {
+      Entity user = datastore.get((new User(id)).getKey());
+      return (String) user.getProperty("email");
```

Comment:
When datastore is unable to find the user, it will return null. This code snippet needs to be able to handle this case or else Java will throw a NullPointerException on line 109 and show the error to the user.

Reply:
Actually never mind, looks like the try/catch deals with it.